lang/LangPrimSource/SC_Ableton_Link.cpp

```diff
@@ -0,0 +1,283 @@
+	: TempoClock(inVMGlobals, inTempoClockObj, inTempo, inBaseBeats, inBaseSeconds),
+	mLink(inTempo * 60.)
+{
+	//quantum = beatsPerBar
+	int err = slotDoubleVal(&inTempoClockObj->slots[2], &mQuantum);
+	if(err) throw err;
+	mLatency = 0.;  // default, user should override
+
+	mLink.enable(true);
+	mLink.enableStartStopSync(true);
+	mLink.setTempoCallback([this](double bpm) {
+		double secs = elapsedTime();
+		double tempo = bpm / 60.;
+
+		auto sessionState = mLink.captureAppSessionState();
+		double beats = sessionState.beatAtTime(hrToLinkTime(secs), mQuantum);
+
+		mTempo = tempo;
+		mBeatDur = 1. / tempo;
+		mCondition.notify_one();
+
+		//call sclang callback
+		gLangMutex.lock();
+		g->canCallOS = false;
+		++g->sp;
+		SetObject(g->sp, mTempoClockObj);
+		++g->sp;
+		SetFloat(g->sp, mTempo);
+		++g->sp;
+		SetFloat(g->sp, beats);
+		++g->sp;
+		SetFloat(g->sp, secs);
+		++g->sp;
+		SetObject(g->sp, mTempoClockObj);
+		runInterpreter(g, getsym("prTempoChanged"), 5);
+		g->canCallOS = false;
+		gLangMutex.unlock();
+	});
+
+	mLink.setStartStopCallback([this](bool isPlaying) {
+		//call sclang callback
+		gLangMutex.lock();
+		g->canCallOS = false;
+		++g->sp;
+		SetObject(g->sp, mTempoClockObj);
+		++g->sp;
+		SetBool(g->sp, isPlaying);
+		runInterpreter(g, getsym("prStartStopSync"), 2);
+		g->canCallOS = false;
+		gLangMutex.unlock();
+	});
+	
+	mLink.setNumPeersCallback([this](std::size_t numPeers) {
+		//call sclang callback
+		gLangMutex.lock();
+		g->canCallOS = false;
+		++g->sp;
+		SetObject(g->sp, mTempoClockObj);
+		++g->sp;
+		SetInt(g->sp, numPeers);
+		runInterpreter(g, getsym("prNumPeersChanged"), 2);
+		g->canCallOS = false;
+		gLangMutex.unlock();
+	});
+	
+	auto sessionState = mLink.captureAppSessionState();
+	auto linkTime = hrToLinkTime(inBaseSeconds);
+	sessionState.requestBeatAtTime(inBaseBeats, linkTime, mQuantum);
+	mLink.commitAppSessionState(sessionState);
+}
+
+void LinkClock::SetAll(double inTempo, double inBeats, double inSeconds)
+{
+
+	auto sessionState = mLink.captureAppSessionState();
+	auto linkTime = hrToLinkTime(inSeconds);
+	sessionState.setTempo(inTempo * 60., linkTime);
+	sessionState.requestBeatAtTime(inBeats,linkTime, mQuantum);
+
+	mTempo = inTempo;
+	mBeatDur = 1. / inTempo;
+
+	mLink.commitAppSessionState(sessionState);
+	mCondition.notify_one();
+}
+
+void LinkClock::SetTempoAtBeat(double inTempo, double inBeats)
+{
+	auto sessionState = mLink.captureAppSessionState();
+	auto time = sessionState.timeAtBeat(inBeats, mQuantum);
+	sessionState.setTempo(inTempo*60., time);
+
+	mTempo = inTempo;
+	mBeatDur = 1. / inTempo;
+
+	mLink.commitAppSessionState(sessionState);
+	mCondition.notify_one();
+}
+
+void LinkClock::SetTempoAtTime(double inTempo, double inSeconds)
+{
+	auto sessionState = mLink.captureAppSessionState();
+	sessionState.setTempo(inTempo*60., hrToLinkTime(inSeconds));
+
+	mTempo = inTempo;
+	mBeatDur = 1. / inTempo;
+
+	mLink.commitAppSessionState(sessionState);
+	mCondition.notify_one();
```

Comment:
`SetAll`, `SetTempoAtBeat`, and `SetTempoAtTime` can be factored -- `CommitTempo(sessionState, tempo)`

Reply:
OK.